phygnn/layers/custom_layers.py

```diff
@@ -943,6 +941,89 @@ def call(self, x):
         return output
 
 
+class AxialAttentionBlock(tf.keras.layers.Layer):
+    """Axial Self-Attention block
+
+    References
+    ----------
+    1. Axial Attention in Multidimensional Transformers.
+       https://arxiv.org/abs/1912.12180
+    """
+
+    def __init__(self, num_heads=4, key_dim=64, name=None):
+        """
+        Parameters
+        ----------
+        num_heads : int
+            Number of attention heads
+        key_dim : int
+            Size of each attention head
+        name : str | None
+            Name of layer.
+        """
+
+        super().__init__(name=name)
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.attention_layers = None
+        self.rank = None
+
+    def build(self, input_shape):
+        """Build the AxialAttentionBlock layer based on an input shape
+
+        Parameters
+        ----------
+        input_shape : tuple
+            Shape tuple of the input tensor
+        """
+        self.rank = len(input_shape)
+        self.attention_layers = [
+            tf.keras.layers.MultiHeadAttention(
+                num_heads=self.num_heads, key_dim=self.key_dim
+            )
+            for _ in range(self.rank - 2)
+        ]
+
+    def _apply_attention_along_axis(self, x, axis):
+        """Apply attention along the given axis. Flattens all other
+        spatiotemporal axes and then reshapes the attended output"""
+
+        # Permute so the axis becomes the "sequence" dimension
+        order = list(range(self.rank))
+        order[axis], order[-2] = order[-2], order[axis]
+        x_perm = tf.transpose(x, perm=order)
+
+        # Flatten the non-sequence dims and compute attention
+        x_flat = tf.reshape(x_perm, (-1, x_perm.shape[-2], x_perm.shape[-1]))
+        x_attn = self.attention_layers[axis - 1](x_flat, x_flat)
+        x_attn = tf.keras.layers.Add()([x_flat, x_attn])  # Residual
+
+        # Reshape and permute back
+        x_attn = tf.reshape(x_attn, x_perm.shape)
+        inv_permute = tf.argsort(order)
+        x_attn = tf.transpose(x_attn, perm=inv_permute)
+        return x_attn
+
+    def call(self, x):
+        """Applies attention along each spatial and the temporal axis
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            Input tensor.
+
+        Returns
+        -------
+        x : tf.Tensor
+            Output tensor
+        """
+        x = self._apply_attention_along_axis(x, axis=1)  # Lats
+        x = self._apply_attention_along_axis(x, axis=2)  # Lons
+        if self.rank == 5:
+            x = self._apply_attention_along_axis(x, axis=3)  # Time
+        return x
```

Comment:
Recommend iterating `for idx, ida in enumerate(range(1, x.ndim-1))` and passing `self.attention_layers[idx]` to the apply method. Much more flexible than assuming rank==4/5 which isn't currently documented. If you do this, note somewhere that we assume ida=0 is obs and ida=-1 is features.

Reply:
Good suggestion, thanks. Think we can just pass ida to the current apply method though?

Reply:
Sure, you get my point though right? It would be nice if this was flexible enough to run on a 3D tensor with timeseries without mods.

Reply:
Yeah, doesn't just passing ida the current method do that? I don't follow the suggestion to pass the layer through the method.

Reply:
Yeah your new commit resolves this comment, just making sure we're on the same page on flexibility